compiler/modules/types/list.bal

```diff
@@ -572,13 +572,60 @@ function listSubtypeIntersect(Context cx, SubtypeData t1, SubtypeData t2) return
 }
 
 function listSubtypeDiff(Context cx, SubtypeData t1, SubtypeData t2) returns SubtypeData {
+    MemoBddCache cache = new (cx.listMemo);
+    if listIsEmptySimple(cx, bddIntersect(cache, <Bdd> t1, <Bdd> t2)) is true {
+        return t1;
+    }
     Bdd b = memoSubtypeDiff(cx.listMemo, <Bdd>t1, <Bdd>t2);
     if b != false && cx.listMemo[b] == () {
         //io:println("fresh BDD: ", bddToString(b));
     }
     return b;
 }
 
+function listIsEmptySimple(Context cx, Bdd bdd) returns boolean? {
+    BddMemo? m = cx.listMemo[bdd];
+    if m !is () && m.empty is boolean {
+        return <boolean>m.empty;
+    }
+    boolean empty = bddEvery(cx, bdd, (), (), listFormulaIsDefinitelyEmpty);
+    if empty == true {
+        cx.listMemo.put({ bdd, empty }); // handle both memoized values being () and not in memo
+        return true;
+    }
+    return ();
+}
+
+function listFormulaIsDefinitelyEmpty(Context cx, Conjunction? pos, Conjunction? neg) returns boolean {
+    if neg !is () || pos is () {
+        return false;
+    }
+    var { members, rest } = cx.listAtomType(pos.atom);
+    int memberCount = members.fixedLength;
+    if !isNever(rest) {
```

Comment:
We should figure out the right requirement on `rest`.

I have `S = [S1, Sr...]`, and `T = [T1, Tr...]` and `S1&T1` is never, then `S&T` is empty.

Reply:
fixed. (I think as long as any member of the intersection is never intersection is empty. Originally I had this check to limit this to tuples only)